File: LeetCode/remote/decode_ways.cs
//  https://leetcode.com/problems/decode-ways/
//
//   A message containing letters from A-Z is being encoded to numbers using the following mapping:
//
//   'A' -> 1
//   'B' -> 2
//   ...
//   'Z' -> 26
//
//   Given an encoded message containing digits, determine the total number of ways to decode it.
//
//   For example,
//   Given encoded message "12", it could be decoded as "AB" (1 2) or "L" (12).
//
//   The number of ways decoding "12" is 2. 

using System;
using System.Collections.Generic;

public class Solution {
    //  
    //  Submission Details
    //  259 / 259 test cases passed.
    //      Status: Accepted
    //      Runtime: 144 ms
    //          
    //          Submitted: 0 minutes ago
    //
    //  Yay! Did this all by myself after fishing out all the fucking corner-cases from the OJ
    //  Remember kids, CORNER CASES!!
    public int NumDecodings(string s) {
        var dp = new int[s.Length + 1];
        if (String.IsNullOrEmpty(s))
        {
            return 0;
        }

        dp[s.Length] = 1;
        for (var i = s.Length - 1; i >= 0; i--)
        {
            if (s[i] == '0')
            {
                dp[i] = 0;
            }
            else
            {
                var nextTwo = i + 1 < s.Length ? int.Parse(s.Substring(i, 2)) : 0;
                dp[i] = dp[i + 1] + (0 < nextTwo && nextTwo < 27 ? dp[i + 2] : 0);
            }
        }

        return dp[0];
    }

    //  https://leetcode.com/submissions/detail/64023392/
    //
    //  Submission Details
    //  259 / 259 test cases passed.
    //      Status: Accepted
    //      Runtime: 128 ms
    //          
    //          Submitted: 0 minutes ago
    public int NumDecodings(string s) {
        if (String.IsNullOrWhiteSpace(s)) {
            return 0;
        }
        
        var ways = new int[s.Length + 1];
        ways[s.Length] = 1;
        
        for (var i = s.Length - 1; i >= 0; i--) {
            if (s[i] == '0') {
                continue;
            }
            
            ways[i] = ways[i + 1];
            
            if (i < s.Length - 1 && int.Parse(s.Substring(i, 2)) < 27)
            {
                ways[i] += ways[i + 2];
            }
        }
        
        return ways[0];
    }

    static void Main()
    {
        var s = new Solution();
        // 1 2 2 6
        // 12 2 6
        // 12 26
        // 1 2 26
        // 1 22 6
        foreach (var t in new [] { "12", "1226", "", "1", "00", "01" })
        {
            Console.WriteLine("{0} - {1}", t, s.NumDecodings(t));
        }
    }
}

File: LeetCode/bootleg/decode_ways.rb
def num_decodings(s)
    return s.nil? || s.empty? ? 
                1 :
                (s[0] == 0 ? 0 : num_decodings(s[1..-1])) + 
                (s.size > 1 && s[0, 2].to_i < 27 ? num_decodings(s[3..-1]) : 0)
end

class String
    #   
    #   Submission Details
    #   259 / 259 test cases passed.
    #       Status: Accepted
    #       Runtime: 84 ms
    #           
    #           Submitted: 0 minutes ago
    #
    #   https://leetcode.com/submissions/detail/51886086/
    #
    #   You are here!
    #   Your runtime beats 100.00% of ruby submissions.
    #
    #   YEah!! Really, there are probably like 2 other submissions...
    def num_decodings
        s = self
        return 0 if s.nil? || s.empty?
        dp = Array.new(s.size + 1)
        dp[s.size] = 1
        (s.size - 1).downto(0).each { |i|
            dp[i] = 0
            next if s[i] == '0'
            dp[i] = dp[i + 1] + (i < s.size - 1 && s[i, 2].to_i < 27 ? dp[i + 2] : 0)
        }

        dp[0]
    end

    #
    #   Submission Details
    #   259 / 259 test cases passed.
    #       Status: Accepted
    #       Runtime: 80 ms
    #           
    #           Submitted: 0 minutes ago
    #
    #   https://leetcode.com/submissions/detail/51890817/
    #
    #   You are here!
    #   Your runtime beats 100.00% of ruby submissions.
    def num_decode
        s = self
        return 0 if s.nil? || s.empty?
        dp = Array.new(s.size + 1)
        dp[0] = 1
        1.upto(s.size).each { |i|
            dp[i] = s[i - 1] == "0" ? 0 : dp[i - 1]
            dp[i] += i > 1 && s[i - 2] != "0" && s[i - 2, 2].to_i < 27 ? dp[i - 2] : 0
        }

        dp[s.size]
    end
end

puts num_decodings("12")
["10", "12"].each { |i|
    puts "#{i} #{i.num_decodings} #{i.num_decode}"
}
